### PR TITLE
3d vines

### PR DESCRIPTION
**Pull request by Razharas**
Your branch is behind i assume concidering all the commits

### CODE REVIEW

**Comment by Razharas** [comment]
I hope that all this nonsense is just your pr being behind upstream master

**Comment by RemieRichards** [comment]
A little behind, yes lol.

File: code/modules/events/spacevine.dm [hunk 1]
```diff
@@ -554,24 +560,24 @@
...
+	var/direction = pick(dirs)
 	var/turf/stepturf = get_step(src,direction)
-	for(var/datum/spacevine_mutation/SM in mutations)
-		SM.on_spread(src, stepturf)
-		stepturf = get_step(src,direction) //in case turf changes, to make sure no runtimes happen
-	if(!locate(/obj/structure/spacevine, stepturf))
-		if(stepturf.Enter(src))
-			if(master)
-				master.spawn_spacevine_piece(stepturf, src)
+	if(stepturf)
```
Comment (RemieRichards) [inline code/modules/events/spacevine.dm:588]: You need to use the helpers, to see if the Z levels are connected (see the helpers file in the multiz folder)

File: code/modules/events/spacevine.dm [hunk 2]
```diff
@@ -554,24 +560,24 @@
...
+		for(var/datum/spacevine_mutation/SM in mutations)
+			SM.on_spread(src, stepturf)
+			stepturf = get_step(src,direction) //in case turf changes, to make sure no runtimes happen
```
Comment (RemieRichards) [inline code/modules/events/spacevine.dm:591]: Same again.

**Comment by Razharas** [comment]
@RemieRichards made the zlevel check on the init and movement(unlikely situation but still)
Mutation specific checks are still there coz for example bluespace doesnt give a fuck if a turf is z_open or not on spread

**Comment by Razharas** [comment]
This ofcourse assumes zlevel configuration wont change at runtime

**Comment by RemieRichards** [comment]
Connected zlevels won't change, no, if at roundstart 5 isn't connected to 6, they never will be (besides badminnery)

Also bluespace probably should still check, otherwise it's going to (logically) appear randomly in the middle of places they shouldn't be (if two zs aren't connected, they should be considered infinitely apart from each other (or horizontal, etc.))

**Comment by RemieRichards** [comment]
Also you should check the turfs are z_open, otherwise you'll have vines growing THROUGH the floor/ceiling.
there's helpers for this, ```GetZOpenAbove()```/```GetZOpenBelow()```

**Comment by Razharas** [comment]
Bluespace(like everything else) checks the connectivity of zlevels but doesnt care if the turf above or below passable or not(all other viens care and work as if these turfs dont exist)

**Comment by RemieRichards** [comment]
<del>Are you sure? it looks like NONE of them check if it's passable or not (just connected)</del>
Actually yeah, they seem right

**Comment by Razharas** [comment]
Well unless there is no check for z_open on Entered(all vines work by checking it) then it should work

**Comment by Razharas** [comment]
Nice

**Comment by RemieRichards** [comment]
:shipit:

**Pull request merged**